compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmOptimizationLowering.kt

```diff
@@ -71,9 +72,38 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
         }
 
     override fun lower(irFile: IrFile) {
-        irFile.transformChildrenVoid(object : IrElementTransformerVoid() {
-            override fun visitCall(expression: IrCall): IrExpression {
-                expression.transformChildrenVoid(this)
+        val transformer = object : IrElementTransformer<IrClass?> {
+
+            // Thread the current class through the transformations in order to replace
+            // final default accessor calls with direct backing field access when
+            // possible.
+            override fun visitClass(declaration: IrClass, data: IrClass?): IrStatement {
+                declaration.transformChildren(this, declaration)
+                return declaration
+            }
+
+            // For some functions, we clear the current class field since the code could end up
+            // in another class then the one it is nested under in the IR.
+            // TODO: Loosen this up for local functions for lambdas passed as an inline lambda
+            // argument to an inline function. In that case the code does end up in the current class.
+            override fun visitFunction(declaration: IrFunction, currentClass: IrClass?): IrStatement {
+                val codeMightBeGeneratedInDifferentClass = declaration.isSuspend ||
+                        declaration.isInline ||
+                        declaration.origin == IrDeclarationOrigin.LOCAL_FUNCTION_FOR_LAMBDA
+                declaration.transformChildren(this, if (codeMightBeGeneratedInDifferentClass) null else currentClass)
```

Comment:
Nit: good place for `takeIf`: `currentClass.takeIf { codeMightBeGeneratedInDifferentClass }`

Reply:
`takeUnless` probably

Reply:
Very likely! :) Thank you!

Reply:
Done. Thank you. :)